src/client/app/utils/Datasources.js

```diff
@@ -0,0 +1,34 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```

Comment:
So this is where I'm going to really talk about the datasource stuff that I've commented about everywhere in depth.
Sorry if it's repetitive; the pull request is really long and I didn't realize I wanted to centralize this comment until I had commented in a bunch of other places.

Anyway, I think that the idea of a datasource is fundamentally a bad abstraction. It looks like we wanted to refactor things to be more abstract and to use `MultiSelectComponent`, and that we decided to abstract away meters and groups into a single "DataSource" to make things easier there. The problem, though, is that we don't want to abstract meters and groups because every place in the code that uses them deals with them differently. The result is code that looks like this (although the meter- or group-specific code isn't often refactored out to its own method):

```
function doThingWithDatasources(dsIDs, dstype) {
	if (dstype === DATA_TYPE_METER) {
		doThingWithMeterIDs(dsIDs);
	else if (dstype === DATA_TYPE_GROUP) {
		doThingWithGroupIDs(dsIDs);
	else {
		// throw an error, we don't know about this dstype
	}
}
```

This is the wrong abstraction. There is no code shared between the `dstype === DATA_TYPE_METER` and `dstype === DATA_TYPE_GROUP` branches because we aren't operating on a higher level of abstraction that meters and groups share, since that higher level of abstraction does not exist.

Instead, we should do this:

```
function doThingWithMeterIDs(meterIDs) {
	// do stuff with meter ids
}

function doThingWithGroupIDs(groupIDs) {
	// do stuff with group ids
}
```

Or, if we want to do something with both at the same time,

```
function doSomethingWithMeterAndGroupIDs(meterIDs, groupIDs) {
	doThingWithMeterIDs(meterIDs);
	doThingWithGroupIDs(groupIDs);
	// Do stuff that requires both of them at the same time.

	// Or, if we're doing something asynchronous,

	return Promise.all([doThingWithMeterIDs(meterIDs), doThingWithGroupIDs(groupIDs)])
		.then([meterResult, groupResult] => {
			// Do stuff after both are done, like dispatching redux actions.
		});
}
```

[Here][1]'s an interesting blog post about the concept of "the wrong abstraction". It concentrates on what happens after an abstraction that seems to work well breaks down as more special-cases are added, but in our case, the abstraction has already broken down with just the two cases of meters and groups. The blog post describes the bad-abstraction scenario as follows (emphasis mine): "The code no longer represents a single, common abstraction, but has instead become **a condition-laden procedure which interleaves a number of vaguely associated ideas**. It is hard to understand and easy to break." That's where we are now.


Fortuantely, this is a pretty easy fix and I think it will actually reduce the amount of code in the project. We just go back to passing around meters and groups seperately and dealing with as what they are.

[1]: https://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction

Reply:
I agree with you on this. I initially began working in the opposite direction - a robust OO approach, where datasources knew how to do most common actions themselves - and implemented this as a "lite" version of that which I could easily finish in the limited time available before AASHE. I'd be happy to go with either fully separated implementations or the OO approach I was originally working on, but this solution isn't a good long term one.

Reply:
My problem with going for an OO-approach is that it might make our redux store much more complicated, since you can only use a string (and therefore sort of an integer) as an object key. If you have a good solution to that then OO is worth looking in to, otherwise I'd say we should keep it simple.